TESTS/mbed_platform/FileHandle/TestFile.h

```diff
@@ -0,0 +1,142 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_TESTFILEHANDLE_H
+#define MBED_TESTFILEHANDLE_H
+
+#include "platform/FileHandle.h"
+
+
+#define POS_IS_VALID(pos) (pos >= 0 && pos < _end)
+#define NEW_POS_IS_VALID(pos) (pos >= 0 && pos < (int32_t)FILE_SIZE)
+#define SEEK_POS_IS_VALID(pos) (pos >= 0 && pos <= _end)
+#define INVALID_POS (-1)
+
+template<uint32_t FILE_SIZE>
+class TestFile : public FileHandle {
+public:
+    TestFile(): _pos(0), _end(0) {}
+    ~TestFile() {}
+
+    enum FunctionName {
+        fnNone, fnRead, fnWrite, fnSeek, fnClose, fnIsatty
+    };
+
+    virtual ssize_t read(void *buffer, size_t size)
+    {
+        ssize_t read;
+        _fnCalled = fnRead;
+
+        for(read = 0; (size_t)read < size; read++)
+        {
+            if(POS_IS_VALID(_pos)) {
+                ((uint8_t*)buffer)[read] = _data[_pos++];
+            } else {
+                break;
+            }
+        } // for
+        return read;
+    }
+
+    virtual ssize_t write(const void *buffer, size_t size)
+    {
+        ssize_t written;
+        _fnCalled = fnWrite;
+
+        for(written = 0; (size_t)written < size; written++)
+        {
+            if(NEW_POS_IS_VALID(_pos)) {
+                _data[_pos++] = ((uint8_t*)buffer)[written];
+            } else {
+                break;
```

Comment:
I don't think this works for write - it's not symmetrical with read. You can't just stop writing data and start returning 0 - there's no "end of file" when writing. Not sure what the C libraries will do if you do this.

While in theory you can do what you like with a FileHandle (or POSIX file descriptor), the C library will be assuming somewhat conventional file-like behaviour, and that doesn't include just not making progress. The C library may well keep continuously recalling you until you return an error code, or the total amount had been written.

You will need a specific error code like `-ENOSPC` here to indicate that they're writing beyond the capacity, and to make sure the C library gives up.

So I think it's return 0 if size == 0, return the amount actually written if you had to stop due to lack of space, and return -ENOSPC if can't write anything due to lack of space.

Reply:
fixed